src/H5Dmpio.c

```diff
@@ -2954,464 +3398,1569 @@ H5D__chunk_redistribute_shared_chunks(const H5D_io_info_t *io_info, const H5D_ty
+    /* Perform all the recvs on the chunks this rank DOES own */
+    for (i = 0, last_assigned_idx = 0; i < *chunk_list_num_entries; i++) {
+        H5D_filtered_collective_io_info_t *chunk_entry = &chunk_list[i];
 
-        if (mpi_rank == chunk_entry->owners.new_owner) {
-            /* Allocate all necessary buffers for an asynchronous receive operation */
+        if (mpi_rank == chunk_entry->new_owner) {
             if (chunk_entry->num_writers > 1) {
                 MPI_Message message;
                 MPI_Status  status;
                 size_t      j;
 
-                chunk_entry->async_info.num_receive_requests = (int)chunk_entry->num_writers - 1;
-                if (NULL == (chunk_entry->async_info.receive_requests_array = (MPI_Request *)H5MM_malloc(
+                /* Allocate all necessary buffers for a non-blocking receive operation */
+                chunk_entry->async_info.num_receive_requests = chunk_entry->num_writers - 1;
+                if (NULL == (chunk_entry->async_info.receive_requests_array = H5MM_malloc(
                                  (size_t)chunk_entry->async_info.num_receive_requests * sizeof(MPI_Request))))
-                    HGOTO_ERROR(H5E_DATASET, H5E_CANTALLOC, FAIL, "unable to allocate async requests array")
+                    HGOTO_ERROR(H5E_DATASET, H5E_CANTALLOC, FAIL, "unable to allocate MPI requests array")
 
                 if (NULL ==
-                    (chunk_entry->async_info.receive_buffer_array = (unsigned char **)H5MM_malloc(
+                    (chunk_entry->async_info.receive_buffer_array = H5MM_malloc(
                          (size_t)chunk_entry->async_info.num_receive_requests * sizeof(unsigned char *))))
-                    HGOTO_ERROR(H5E_DATASET, H5E_CANTALLOC, FAIL, "unable to allocate async receive buffers")
+                    HGOTO_ERROR(H5E_DATASET, H5E_CANTALLOC, FAIL, "unable to allocate receive buffers")
 
-                for (j = 0; j < chunk_entry->num_writers - 1; j++) {
+                for (j = 0; j < (size_t)chunk_entry->num_writers - 1; j++) {
                     int count = 0;
 
-                    /* Probe for a particular message from any process, removing that message
-                     * from the receive queue in the process and allocating that much memory
-                     * for the asynchronous receive
+                    /* Probe for a particular message for this chunk from any rank, removing
+                     * that message from the receive queue for the rank and allocating that
+                     * much memory for the non-blocking receive
                      */
-                    if (MPI_SUCCESS != (mpi_code = MPI_Mprobe(MPI_ANY_SOURCE, (int)chunk_entry->index,
-                                                              io_info->comm, &message, &status)))
+                    if (MPI_SUCCESS !=
+                        (mpi_code = MPI_Mprobe(MPI_ANY_SOURCE, (int)chunk_entry->index_info.chunk_idx,
```

Comment:
Here I think the changes are mostly just formatting and struct field renaming, so this blocking probe approach was the old first attempt at getting this correct. This has since been re-written into a non-blocking probing approach (not in this current PR), but I think in general it can be hard to guess about performance since we might be entering MPI multiple times in a non-blocking probing loop as opposed to blocking here. Hard to say, but I think initial profiling with the newly-rewritten non-blocking way will give some ideas.